File: src/main/java/org/distril/beengine/world/chunk/bitarray/PaddedBitArray.java
package org.distril.beengine.world.chunk.bitarray;

import com.google.common.base.Preconditions;

public class PaddedBitArray extends BitArray {

	public PaddedBitArray(Version version, int size, int[] words) {
		super(version, size, words);
	}

	@Override
	public void set(int index, int value) {
		Preconditions.checkElementIndex(index, this.size);
		Preconditions.checkArgument(value >= 0 && value <= this.version.getMaxEntryValue(), "Max value: %s. Received value", this.version.getMaxEntryValue(), value);

		var arrayIndex = index / this.version.getEntriesPerWord();
		var offset = (index % this.version.getEntriesPerWord()) * this.version.getBits();
		this.words[arrayIndex] = this.words[arrayIndex] & ~(this.version.getMaxEntryValue() << offset) | (value & this.version.getMaxEntryValue()) << offset;
	}

	@Override
	public int get(int index) {
		Preconditions.checkElementIndex(index, this.size);

		var arrayIndex = index / this.version.getEntriesPerWord();
		var offset = (index % this.version.getEntriesPerWord()) * this.version.getBits();
		return (this.words[arrayIndex] >>> offset) & this.version.getMaxEntryValue();
	}
}

File: src/main/kotlin/org/distril/beengine/world/chunk/bitarray/PaddedBitArray.kt
package org.distril.beengine.world.chunk.bitarray

import org.distril.beengine.util.Utils.requireInRange

class PaddedBitArray(version: Version, size: Int, words: IntArray) : BitArray(version, size, words) {

	override fun set(index: Int, value: Int) {
		requireInRange(index, maxValue = this.size, name = "index")
		require(value >= 0 && value <= this.version.maxEntryValue) {
			"Max value: ${this.version.maxEntryValue}. Received value $value"
		}

		val arrayIndex = index / this.version.entriesPerWord
		val offset = index % this.version.entriesPerWord * this.version.bits
		this.words[arrayIndex] = this.words[arrayIndex] and
				(this.version.maxEntryValue shl offset).inv() or
				(value and this.version.maxEntryValue shl offset)
	}

	override fun get(index: Int): Int {
		requireInRange(index, maxValue = this.size, name = "index")

		val arrayIndex = index / this.version.entriesPerWord
		val offset = index % this.version.entriesPerWord * this.version.bits
		return this.words[arrayIndex] ushr offset and this.version.maxEntryValue
	}
}
